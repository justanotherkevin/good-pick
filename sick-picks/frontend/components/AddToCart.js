import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { CURRENT_USER_QUERY } from './User';

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`;

class AddToCart extends Component {
  render() {
    const { id } = this.props;
    return (
      <Mutation
        mutation={ADD_TO_CART_MUTATION}
        variables={{ id }}
        refetchQueries={[{ query: CURRENT_USER_QUERY }]}
      >
        {/* These props themselves provide information about the state of the network request: (loading, error, data) */}
        {(addToCart, props) => (
          <button disabled={props.loading} onClick={addToCart}>
            Add{props.loading && 'ing'} To Cart ðŸ›’
          </button>
        )}
      </Mutation>
    );
  }
}
export default AddToCart;
