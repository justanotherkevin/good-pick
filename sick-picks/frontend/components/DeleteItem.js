import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from '../gql/query';
import { DELETE_ITEM_MUTATION } from '../gql/mutation';

class DeleteItem extends Component {
  // manually update the cache on the client, so it matches the server
  update = (cache, payload) => {
    // 1. Read the cache for the items we want
    // 2. Filter the deleted item out of cache
    // 3. Rewrite the cache
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    console.log(data)
    data.items = data.items.filter(
      item => item.id !== payload.data.deleteItem.id
    );
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
  };
  render() {
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
      >
        {(deleteItem, { error }) => (
          <button
            onClick={() => {
              if (confirm('Are you sure you want to delete this item?')) {
                deleteItem().catch(err => {
                  alert(err.message);
                });
              }
            }}
          >
            {this.props.children}
          </button>
        )}
      </Mutation>
    );
  }
}

export default DeleteItem;
