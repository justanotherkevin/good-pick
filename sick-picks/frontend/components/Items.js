import React, { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import Item from './Item';
import Pagination from './Pagination';
import { perPage } from '../config';

const ALL_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}) {
    items(first: $first, skip: $skip, orderBy: createdAt_DESC) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;
const ItemsList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  max-width: ${props => props.theme.maxWidth};
  margin: 0 auto;
`;
const Center = styled.div`
  text-align: center;
`;

class Items extends Component {
  render() {
    return (
      <Center>
        <Pagination page={this.props.page} />
        <Query
          query={ALL_ITEMS_QUERY}
          variables={{
            skip: this.props.page * perPage - perPage
          }}
        >
          {/* query have a payload variable that is return; using destructuring to take only data, error,loading */}
          {({ data, error, loading }) => {
            if (loading) return <p>Loading...</p>;
            if (error) return <p>Error: {error.message}</p>;
            return (
              <ItemsList>
                {data.items.map(item => (
                  <Item item={item} key={item.id} />
                ))}
              </ItemsList>
            );
          }}
        </Query>
      </Center>
    );
  }
}
export default Items;
export { ALL_ITEMS_QUERY };
