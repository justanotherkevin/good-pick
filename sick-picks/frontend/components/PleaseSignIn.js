import { Query } from 'react-apollo';
import { CURRENT_USER_QUERY } from '../gql/query';
import styled from 'styled-components';
import Signin from './Signin';

const SignInContainer = styled.div`
  max-width:  400px;
  width: 100%;
  margin: 0 auto;
`;

const PleaseSignIn = props => (
  <Query query={CURRENT_USER_QUERY}>
    {({ data, loading }) => {
      if (loading) return <p>Loading...</p>;
      if (!data.me) {
        return (
          <SignInContainer>
            <p>Please Sign In before Continuing</p>
            <Signin />
          </SignInContainer>
        );
      }
      return props.children;
    }}
  </Query>
);

export default PleaseSignIn;
