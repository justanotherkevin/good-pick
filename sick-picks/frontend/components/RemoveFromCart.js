import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import gql from 'graphql-tag';
import { Mutation } from 'react-apollo';
import { CURRENT_USER_QUERY } from './User';

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${props => props.theme.red};
    cursor: pointer;
  }
`;
const REMOVE_FROM_CART_MUTATION = gql`
  mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`;
class RemoveFromCart extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired
  };
  update = (cache, payload) => {
    // This gets called as soon as we get a response back from the server after a mutation has been performed
    console.log('Running remove from cart update fn');
    // 1. first read the cache
    const data = cache.readQuery({ query: CURRENT_USER_QUERY });
    console.log(data);
    // 2. remove that item from the cart
    const cartItemId = payload.data.removeFromCart.id;
    data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId);
    // 3. write it back to the cache
    cache.writeQuery({ query: CURRENT_USER_QUERY, data });
  };
  render() {
    return (
      <Mutation
        mutation={REMOVE_FROM_CART_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
        optimisticResponse={{
          __typename: 'Mutation',
          removeFromCart: {
            __typename: 'CartItem',
            id: this.props.id
          }
        }}
      >
        {(removeFromCart, { loading, error, dada }) => (
          <BigButton
            disabled={loading}
            title="Delete Item"
            onClick={() => removeFromCart().catch(err => alert(err.message))}
          >
            &times;
          </BigButton>
        )}
      </Mutation>
    );
  }
}
export default RemoveFromCart;
