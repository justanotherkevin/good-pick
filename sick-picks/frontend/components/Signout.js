import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { CURRENT_USER_QUERY } from '../gql/query';
import { SIGN_OUT_MUTATION } from '../gql/mutation';

const Signout = props => (
  <Mutation
    mutation={SIGN_OUT_MUTATION}
    refetchQueries={[{ query: CURRENT_USER_QUERY }]}
  >
    {signout => <button onClick={signout}>Sign Out</button>}
  </Mutation>
);
export default Signout;
