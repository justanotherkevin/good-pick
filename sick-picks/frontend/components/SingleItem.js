import React, { Component } from 'react';
import { Query } from 'react-apollo';
import Error from './ErrorMessage';
import styled from 'styled-components';
import Head from 'next/head';
import { SINGLE_ITEM_QUERY } from '../gql/query';

const SingleItemStyles = styled.div`
  max-width: 600px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  display: grid;
  grid-template-columns: 1fr;
  /* grid-auto-flow: column; */
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
    padding: 2rem;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
  @media only screen and ( min-width: ${prop => prop.theme.mediaScreenMedium} ) {
    max-width: ${props => props.theme.maxWidth};
    grid-template-columns: 1fr 1fr;
  }
`;


class SingleItem extends Component {
  render() {
    const itemObj = { id: this.props.id };
    return (
      <Query query={SINGLE_ITEM_QUERY} variables={itemObj}>
        {apolloRes => {
          const { error, loading, data } = apolloRes;
          if (error) return <Error error={error} />;
          if (loading) return <p>Loading...</p>;
          // client side resolver if no item found.
          if (!data.item) return <p>No Item Found for {this.props.id}</p>;

          const item = data.item;
          return (
            <SingleItemStyles>
              {/* moddify head, title to the name of the item; sideeffect with next.js*/}
              <Head>
                <title>Good Shop | {item.title}</title>
              </Head>
              <img src={item.largeImage} alt={item.title} />
              <div className="details">
                <h2>{item.title}</h2>
                <p>{item.description}</p>
              </div>
            </SingleItemStyles>
          );
        }}
      </Query>
    );
  }
}
export default SingleItem;
