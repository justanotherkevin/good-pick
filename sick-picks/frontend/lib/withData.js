import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include'
        },
        headers
      });
    },
    // local data
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, client) {
            // for some no reason the first param is underscore.
            // 1. get current cartOpen val
            const { cartOpen } = client.cache.readQuery({
              query: LOCAL_STATE_QUERY
            });
            // 2. new vonst for data
            const data = {
              data: { cartOpen: !cartOpen }
            };
            client.cache.writeData(data);
            return data;
          }
        }
      },
      defaults: {
        cartOpen: false
      }
    }
  });
}

export default withApollo(createClient);
