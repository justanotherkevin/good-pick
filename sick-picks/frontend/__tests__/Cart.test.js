import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import { ApolloConsumer } from 'react-apollo';
import { fakeUser, fakeCartItem } from '../lib/testUtils';
import { CURRENT_USER_QUERY } from '../components/User';
import Cart, { LOCAL_STATE_QUERY } from '../components/Cart';

const mocks = [
  {
    request: {
      query: CURRENT_USER_QUERY
    },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem()]
        }
      }
    }
  },
  {
    request: { query: LOCAL_STATE_QUERY },
    result: {
      data: {
        cartOpen: true
      }
    }
  }
];

describe('<RequestReset/> component test', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Cart />
      </MockedProvider>
    );
    await wait();
    expect(toJSON(wrapper.find('header'))).toMatchSnapshot();
    // console.log(form.debug());
  });
});
