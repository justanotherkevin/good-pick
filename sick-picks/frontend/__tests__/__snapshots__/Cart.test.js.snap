// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset/> component test renders and matches snapshot 1`] = `null`;
